File: Data Structures & Algorithms/binary-tree-diameter/submission-2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun diameterOfBinaryTree(root: TreeNode?): Int {
        var ans = Int.MIN_VALUE
        fun maxDepth(root: TreeNode?): Int {        
        if(root == null) {
            return 0
        }
        val l = maxDepth(root.left)
        val r = maxDepth(root.right)

        ans = maxOf(ans, l + r)
        return 1+ maxOf(l, r)
    }
    maxDepth(root)
    return ans
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        map = dict()
        l, ans = 0,0
        for i in range(len(s)):
            if s[i] in map:
                l = max(l, map[s[i]])
            ans = max(ans, i - l + 1)
            map[s[i]] = i + 1
        return ans